### PR TITLE
[VER-1906] Add age and gender custom dimensions to GA events where available

### PR DESCRIPTION
**Pull request by cjrudd**

### CODE REVIEW

File: app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala [hunk 1]
```diff
@@ -46,16 +59,49 @@ object PersonalDetails {
...
+
 case class PersonalDetailsWithNino(firstName: NonEmptyString,
                                    lastName: NonEmptyString,
                                    dateOfBirth: LocalDate,
                                    nino: Nino) extends PersonalDetails with PersonalDetailsNino {
+
+  override def addGender(gender: NonEmptyString): PersonalDetails =
+    PersonalDetailsWithNinoAndGender(firstName, lastName, dateOfBirth, nino, gender)
+
+  override def maybeNino: Option[Nino] = Some(nino)
```
Comment (stevenmallen) [inline app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala:71]: Seems a bit odd to have a maybeNino method in a class called "PersonalDetailsWithNino", but then I'm not really following all the changes clearly, sorry.

File: app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala [hunk 2]
```diff
@@ -46,16 +59,49 @@ object PersonalDetails {
...
+
+case class PersonalDetailsWithNinoAndGender(firstName: NonEmptyString,
+                                            lastName: NonEmptyString,
+                                            dateOfBirth: LocalDate,
+                                            nino: Nino,
+                                            gender: NonEmptyString) extends PersonalDetails with PersonalDetailsNino with PersonalDetailsGender {
+
+  override def maybeNino: Option[Nino] = Some(nino)
+
+  override def maybeGender: Option[NonEmptyString] = Some(gender)
```
Comment (stevenmallen) [inline app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala:93]: Again, the class name would suggest that Gender is always part of this data model?

File: app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala [hunk 3]
```diff
@@ -46,16 +59,49 @@ object PersonalDetails {
...
+
+  lazy val toJson: JsObject = Json.obj(
+    "firstName" -> firstName,
+    "lastName" -> lastName,
+    "dateOfBirth" -> dateOfBirth,
+    "nino" -> nino,
+    "gender" -> gender
+  )
+
+  def age: Int = YEARS.between(dateOfBirth, LocalDate.now()).toInt
```
Comment (stevenmallen) [inline app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala]: Can/should this common code be refactored into another trait?

File: app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala [hunk 4]
```diff
@@ -72,6 +118,8 @@ case class PersonalDetailsWithPostCode(firstName: NonEmptyString,
     "dateOfBirth" -> dateOfBirth,
     "postCode" -> postCode.value
   )
+
+  def age: Int = YEARS.between(dateOfBirth, LocalDate.now()).toInt
```
Comment (stevenmallen) [inline app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala]: Ditto

File: app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala [hunk 5]
```diff
@@ -82,11 +130,48 @@ case class PersonalDetailsWithNinoAndPostCode(firstName: NonEmptyString,
...
+
   lazy val toJson: JsObject = Json.obj(
     "firstName" -> firstName,
     "lastName" -> lastName,
     "dateOfBirth" -> dateOfBirth,
     "postCode" -> postCode.value,
     "nino" -> nino
   )
+
+  def age: Int = YEARS.between(dateOfBirth, LocalDate.now()).toInt
```
Comment (stevenmallen) [inline app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala]: Ditto

File: app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala [hunk 6]
```diff
@@ -82,11 +130,48 @@ case class PersonalDetailsWithNinoAndPostCode(firstName: NonEmptyString,
...
+  lazy val toJson: JsObject = Json.obj(
+    "firstName" -> firstName,
+    "lastName" -> lastName,
+    "dateOfBirth" -> dateOfBirth,
+    "postCode" -> postCode.value,
+    "nino" -> nino,
+    "gender" -> gender
+  )
+
+  def age: Int = YEARS.between(dateOfBirth, LocalDate.now()).toInt
```
Comment (stevenmallen) [inline app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala]: Ditto

File: app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala (hunk 5 above)
Comment (cjrudd) [reply app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala]: done

File: app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala (hunk 6 above)
Comment (cjrudd) [reply app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala]: done

File: app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala (hunk 4 above)
Comment (cjrudd) [reply app/uk/gov/hmrc/personaldetailsvalidation/model/PersonalDetails.scala]: done

**Review by stevenmallen — APPROVED** [review]
LGTM

**Pull request merged**